libclc/ptx-nvidiacl/libspirv/images/image.cl

```diff
@@ -2512,26 +2512,31 @@ uint4 __nvvm_tex_3d_grad_v4j32_f32(
     unsigned long, float, float, float, float, float, float, float, float,
     float) __asm("__clc_llvm_nvvm_tex_3d_grad_v4j32_f32");
 
-// Macro to generate mipmap vec2 fetches
-#define _CLC_DEFINE_MIPMAP_BINDLESS_VEC2THUNK_READS_BUILTIN(                   \
-    elem_t, dimension, vec_size, fetch_vec_size, coord_input, coord_parameter, \
-    grad_input, ...)                                                           \
-  elem_t##2 __nvvm_tex_##dimension##d_level_##vec_size##_f32(                  \
+// Macro to generate mipmap vec4 fetches
+#define _CLC_DEFINE_MIPMAP_BINDLESS_VEC4THUNK_READS_BUILTIN(                   \
+    elem_t, fetch_elem_t, dimension, vec_size, fetch_vec_size, coord_input,    \
+    coord_parameter, grad_input, ...)                                          \
+  elem_t##4 __nvvm_tex_##dimension##d_level_##vec_size##_f32(                  \
       unsigned long imageHandle, coord_input, float level) {                   \
-    elem_t##4 a = __nvvm_tex_##dimension##d_level_##fetch_vec_size##_f32(      \
-        imageHandle, coord_parameter, level);                                  \
-    elem_t##2 ret;                                                             \
-    ret.x = a.x;                                                               \
-    ret.y = a.y;                                                               \
+    fetch_elem_t##4 a =                                                        \
+        __nvvm_tex_##dimension##d_level_##fetch_vec_size##_f32(                \
+            imageHandle, coord_parameter, level);                              \
+    elem_t##4 ret;                                                             \
+    ret.x = (elem_t)a.x;                                                       \
+    ret.y = (elem_t)a.y;                                                       \
+    ret.z = (elem_t)a.z;                                                       \
+    ret.w = (elem_t)a.w;                                                       \
     return ret;                                                                \
   }                                                                            \
```

Comment:
I remember libclc/ptx-nvidiacl/libspirv/images/image.cl defining casts for vectors too. I wonder if there would be a benefit to consolidating the implementations.

Reply:
Thanks for the feedback! The macro casts are now used instead where possible. Reduced the number of lines by a decent amount.